jcl/src/java.base/share/classes/openj9/internal/foreign/abi/LayoutStrPreprocessor.java

```diff
@@ -209,38 +218,150 @@ private static StringBuilder preprocessLayout(MemoryLayout targetLayout, boolean
+	}
+
+	/*[IF JAVA_SPEC_VERSION >= 21]*/
+	/* Encode the types in the union layout to a symbol string to simplify the further processing in native. */
+	private static StringBuilder encodeUnionLayoutStr(UnionLayout unionLayout, StringBuilder targetLayoutStr, boolean isDownCall) {
+		long unionByteSize = unionLayout.byteSize();
+		StringBuilder elementLayoutStrs = new StringBuilder();
+		int byteAlignment = (int)unionLayout.byteAlignment();
+		ValueLayout sequenceElement;
+
+		/* The idea is to select the primitive type with the biggest alignment of the union
+		 * (obtained via unionLayout.byteAlignment()) to be laid out as a sequence which is
+		 * correctly mappped onto the given GPRs/FPRs on the hardware.
+		 */
+		switch (byteAlignment) {
+		case 1: /* 1 Byte. */
+			sequenceElement = JAVA_BYTE;
+			break;
+		case 2: /* 2 Bytes. */
+			sequenceElement = JAVA_SHORT;
+			break;
+		case 4: /* 4 Bytes. */
+			if (validatePrimTypesOfUnionForGPR(unionLayout, byteAlignment)) {
+				sequenceElement = JAVA_INT;
+			} else {
+				sequenceElement = JAVA_FLOAT;
+			}
+			break;
+		case 8: /* 8 Bytes. */
+			if (validatePrimTypesOfUnionForGPR(unionLayout, byteAlignment)) {
+				sequenceElement = JAVA_LONG;
+			} else {
+				sequenceElement = JAVA_DOUBLE;
 			}
-			targetLayoutString.append('[').append(elementLayoutStrs).append(']');
+			break;
+		default:
+			throw new IllegalArgumentException("Invalid alignment of the union layout: " + unionLayout);
+		}
+
+		long elementByteSize = sequenceElement.byteSize();
+		long sequenceCount = (unionByteSize / elementByteSize)
+				+ ((0 == (unionByteSize % elementByteSize)) ? 0 : 1);
+
+		/* There is no need to create a sequence layout for a single element. */
+		MemoryLayout sequenceLayout = (sequenceCount != 1)
+				? MemoryLayout.sequenceLayout(sequenceCount, sequenceElement)
+				: sequenceElement;
+		elementLayoutStrs.append(preprocessLayout(sequenceLayout, isDownCall));
+
+		/* Prefix "#" to denote the start of this layout string in the case of downcall. */
+		if (isDownCall) {
+			targetLayoutStr.append('#').append(1);
 		}
+		targetLayoutStr.append('[').append(elementLayoutStrs).append(']');
 
-		return targetLayoutString;
+		return targetLayoutStr;
 	}
 
+	/* Check whether the specified 4/8-byte aligned group layout (struct/union) layout contains
+	 * non-floating-point primitives so as to determine whether the elements of union can be safely
+	 * mapped to GPRs.
+	 */
+	private static boolean validatePrimTypesOfUnionForGPR(GroupLayout groupLayout, int byteAlignment) {
+		List<MemoryLayout> elementLayoutList = groupLayout.memberLayouts();
+		boolean arePrimTypesForGPR = false;
+		for (MemoryLayout elementLayout : elementLayoutList) {
+			if (((elementLayout instanceof ValueLayout elemValueLayout) && (validatePrimTypeForGPR(elemValueLayout, byteAlignment)))
+				|| ((elementLayout instanceof GroupLayout elemGroupLayout) && validatePrimTypesOfUnionForGPR(elemGroupLayout, byteAlignment))
```

Comment:
Please remove the redundant parentheses from line 329. I also suggest these lines deserve to be split:
```java
			if (((elementLayout instanceof ValueLayout elemValueLayout)
				&& validatePrimTypeForGPR(elemValueLayout, byteAlignment))
			|| ((elementLayout instanceof GroupLayout elemGroupLayout)
				&& validatePrimTypesOfUnionForGPR(elemGroupLayout, byteAlignment))
			) {
```

Reply:
Agreed and updated against the suggestion above.